src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java

```diff
@@ -363,22 +362,46 @@ else if (abbrStarted) {
     /**
      * Get Abbreviation if it is illegal.
      * @param str name
-     * @param beginIndex begin index
-     * @param endIndex end index
-     * @return true is abbreviation is bigger that required and not in ignore list
+     * @param beginIndex begin index, inclusive
+     * @param endIndex end index, inclusive
+     * @return the abbreviation if bigger than required and not in ignore list
      */
     private String getAbbreviationIfIllegal(String str, int beginIndex, int endIndex) {
         String result = null;
         final int abbrLength = endIndex - beginIndex;
         if (abbrLength > allowedAbbreviationLength) {
-            final String abbr = str.substring(beginIndex, endIndex);
+            final String abbr = getAbbreviation(str, beginIndex, endIndex);
             if (!allowedAbbreviations.contains(abbr)) {
                 result = abbr;
             }
         }
         return result;
     }
 
+    /**
+     * Gets the abbreviation, only including the character at {@code endIndex} if
+     * it is the last character in the string, as it is usually the first capital in
+     * the next word.
+     * <p>
+     * For example, {@code getAbbreviation("getXMLParser", 3, 6)} returns "XML"
+     * (not "XMLP"), and so does {@code getAbbreviation("parseXML", 5, 7)}.
+     * </p>
+     * @param str name
+     * @param beginIndex begin index, inclusive
+     * @param endIndex end index, inclusive
+     * @return the specified abbreviation
+     */
+    private static String getAbbreviation(String str, int beginIndex, int endIndex) {
```

Comment:
This method could be responsible for deciding: given a sequence of upper-case characters from `str.charAt(beginIndex)` to `str.charAt(endIndex)` (inclusive), what is the abbreviation that should be checked for existence in `allowedAbbreviations`?

The original code truncated the final upper-case character so that for example in "getXMLParser" the user could specify "XML" as an allowed abbreviation, and it would truncate "XMLP" to "XML".

Like you said, that can't always work, but it seems like a reasonable assumption... what do you think? Maybe it just needs to be documented so the user knows what to expect?

To your point about handling "single-character abbreviations" where `beginIndex == endIndex` and `allowedAbbreviationLength == -1`, what if this method did the following:

```
// don't truncate final character if it is the last character in the string
// don't truncate final character if there is only a single upper-case character
if (endIndex == str.length() - 1 || beginIndex == endIndex) {
    result = str.substring(beginIndex, endIndex + 1);
}
else {
    result = str.substring(beginIndex, endIndex);
}
```